autogpt_platform/autogpt_libs/autogpt_libs/utils/cache.py

```diff
@@ -1,20 +1,58 @@
+import inspect
 import threading
-from typing import Callable, ParamSpec, TypeVar
+from typing import Any, Awaitable, Callable, ParamSpec, TypeVar, Union, cast, overload
 
 P = ParamSpec("P")
 R = TypeVar("R")
 
 
-def thread_cached(func: Callable[P, R]) -> Callable[P, R]:
+@overload
+def thread_cached(func: Callable[P, Awaitable[R]]) -> Callable[P, Awaitable[R]]: ...
+
+
+@overload
+def thread_cached(func: Callable[P, R]) -> Callable[P, R]: ...
+
+
+def thread_cached(
+    func: Union[Callable[P, R], Callable[P, Awaitable[R]]],
+) -> Union[Callable[P, R], Callable[P, Awaitable[R]]]:
     thread_local = threading.local()
 
-    def wrapper(*args: P.args, **kwargs: P.kwargs) -> R:
-        cache = getattr(thread_local, "cache", None)
-        if cache is None:
-            cache = thread_local.cache = {}
-        key = (args, tuple(sorted(kwargs.items())))
-        if key not in cache:
-            cache[key] = func(*args, **kwargs)
-        return cache[key]
+    if inspect.iscoroutinefunction(func):
+
+        async def async_wrapper(*args: P.args, **kwargs: P.kwargs) -> R:
+            cache = getattr(thread_local, "cache", None)
+            if cache is None:
+                cache = thread_local.cache = {}
+            key = (args, tuple(sorted(kwargs.items())))
```

Comment:
shouldn't this include the name of the function?

```suggestion
            key = (func.__name__, args, tuple(sorted(kwargs.items())))
```

Reply:
No the thread local is local per function

Reply:
It seems you were right: `threading.local()` creates an instance of a thread-specific storage object. I assumed it would be global for the thread, but afaics it isn't.